core/trino-main/src/test/java/io/trino/sql/planner/assertions/JoinMatchPatternBuilder.java

```diff
@@ -0,0 +1,155 @@
+import io.trino.sql.planner.plan.JoinNode;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.node;
+import static io.trino.sql.planner.plan.JoinNode.DistributionType;
+import static io.trino.sql.planner.plan.JoinNode.EquiJoinClause;
+import static io.trino.sql.planner.plan.JoinNode.Type;
+import static io.trino.sql.tree.ComparisonExpression.Operator.EQUAL;
+import static java.util.Objects.requireNonNull;
+
+public class JoinMatchPatternBuilder
+{
+    private Type joinType;
+    private List<ExpectedValueProvider<EquiJoinClause>> equiCriteria;
+    private List<PlanMatchPattern.DynamicFilterPattern> dynamicFilter;
+    private DistributionType distributionType;
+    private Boolean expectedSpillable;
+    private PlanMatchPattern left;
+    private PlanMatchPattern right;
+    private String filter;
+
+    public JoinMatchPatternBuilder withType(Type joinType)
+    {
+        this.joinType = joinType;
+
+        return this;
+    }
+
+    public JoinMatchPatternBuilder withEquiCriteria(List<ExpectedValueProvider<EquiJoinClause>> expectedEquiCriteria)
+    {
+        this.equiCriteria = expectedEquiCriteria;
+
+        return this;
+    }
+
+    public JoinMatchPatternBuilder withEquiCriteria(ExpectedValueProvider<EquiJoinClause> expectedEquiCriteria)
+    {
+        this.equiCriteria = ImmutableList.of(expectedEquiCriteria);
+
+        return this;
+    }
+
+    public JoinMatchPatternBuilder withFilter(String expectedFilter)
+    {
+        this.filter = expectedFilter;
+
+        return this;
+    }
+
+    public JoinMatchPatternBuilder withDynamicFilter(Map<String, String> expectedDynamicFilter)
+    {
+        this.dynamicFilter = expectedDynamicFilter.entrySet().stream()
+                .map(entry -> new PlanMatchPattern.DynamicFilterPattern(entry.getKey(), EQUAL, entry.getValue()))
+                .collect(toImmutableList());
+
+        return this;
+    }
+
+    public JoinMatchPatternBuilder withDynamicFilter(List<PlanMatchPattern.DynamicFilterPattern> expectedDynamicFilter)
+    {
+        this.dynamicFilter = expectedDynamicFilter;
+
+        return this;
+    }
+
+    public JoinMatchPatternBuilder withDistributionType(DistributionType expectedDistributionType)
+    {
+        this.distributionType = expectedDistributionType;
+
+        return this;
+    }
+
+    public JoinMatchPatternBuilder withSpillable(Boolean expectedSpillable)
+    {
+        this.expectedSpillable = expectedSpillable;
+
+        return this;
+    }
+
+    public JoinMatchPatternBuilder withLeft(PlanMatchPattern left)
+    {
+        this.left = left;
+
+        return this;
+    }
+
+    public JoinMatchPatternBuilder withRight(PlanMatchPattern right)
+    {
+        this.right = right;
+
+        return this;
+    }
+
+    public PlanMatchPattern build()
+    {
+        requireNonNull(joinType);
```

Comment:
move it to `withType` (similar for others)

Reply:
Build checks state of builder (if all required parameters are set like type, left, right) — not only if argument we want to set is not null.

Reply:
I mean null check in `withType(arg)` would fail-fast if `null` arg is passed